sphinx/tutorial/rose/furthertopics/upgrading-macro-development.rst

```diff
@@ -0,0 +1,568 @@
+
+.. code-block:: python
+
+   class MyMoreComplexUpgradeMacro(rose.upgrade.MacroUpgrade):
+
+       """Upgrade from 0.3 (Amazing Ama) to 1.0 (Tremendous Trimaran)."""
+
+       BEFORE_TAG = "0.3"
+       AFTER_TAG = "1.0"
+
+       def upgrade(self, config, meta_config=None):
+           """Upgrade the boat!"""
+           # Some code doing something to config goes here.
+           return config, self.reports
+
+We already know how to add an option, so replace
+``# Some code going here...`` with
+``self.add_setting(config, ["namelist:materials", "sail_canvas_sq_m"], "5")``
+
+To perform the check/change in the number of ``misc_branches``, we can
+insert the following lines after the one we just added:
+
+.. code-block:: python
+
+           branch_num = self.get_setting_value(
+                  config, ["namelist:materials", "misc_branches"])
+           if branch_num.isdigit() and float(branch_num) < 6:
+               self.change_setting_value(
+                        config, ["namelist:materials", "misc_branches"], "6")
+
+This extracts the value of ``misc_branches`` (as a string!) and if the
+value represents a positive integer that is less than 6, changes it to
+``"6"``. It's good practice to guard against the possibility that a user
+might have set the value to a non-integer representation like ``'many'``
+- if we don't do this, the macro may crash out when running things like
+``float``.
+
+In a similar way, to flag a warning, insert:
+
+.. code-block:: python
+
+           paddles = self.get_setting_value(
+                          config, ["namelist:materials", "paddling_branches"])
+           if paddles is not None:
+               self.add_report("namelist:materials", "paddling_branches",
+                               paddles, info="Deprecated - probably not needed.",
+                               is_warning=True)
+
+This calls ``self.add_report`` if the option ``paddling_branches`` is
+present. This is a method that notifies the user of actions and issues
+by appending things to the ``self.reports`` list which appears on the
+``return ...line``.
+
+Run ``rose app-upgrade --meta-path=~/rose-meta/ 1.0`` to see the effect of
+your changes. You should see a warning message for
+``namelist:materials=paddling_branches`` as well.
+
+Upgrading Many Versions at Once
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+We've kept in step with the metadata by upgrading incrementally, but
+typically users will need to upgrade across multiple versions. When this
+happens, the relevant macros will be applied in turn, and their changes
+and issues aggregated.
+
+Turn back the clock by reverting your application configuration to look
+like it was at ``0.1``:
+
+.. code-block:: rose
+
+   meta=make-boat/0.1
+
+   [namelist:materials]
+   hollow_tree_trunks=1
+   paddling_twigs=1
+
+Run ``rose app-upgrade --meta-path=~/rose-meta/`` in the application
+directory. You should see the available versions to upgrade to - let's
```

Comment:
I advise clarifying by amending to 'see **that the version has now been downgraded to 0.1 and also see** the available ...' or something to this effect.